// @daohaus/abis/src/lib/abis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOCAL_ABI = void 0;
const tslib_1 = require("tslib");
const baal_json_1 = tslib_1.__importDefault(require("../abi/baal.json"));
const baalSummoner_json_1 = tslib_1.__importDefault(require("../abi/baalSummoner.json"));
const baalAdvTokenSummoner_json_1 = tslib_1.__importDefault(require("../abi/baalAdvTokenSummoner.json"));
const gnosisMultisend_json_1 = tslib_1.__importDefault(require("../abi/gnosisMultisend.json"));
const gnosisProxy_json_1 = tslib_1.__importDefault(require("../abi/gnosisProxy.json"));
const gnosisSignLib_json_1 = tslib_1.__importDefault(require("../abi/gnosisSignLib.json"));
const loot_json_1 = tslib_1.__importDefault(require("../abi/loot.json"));
const poster_json_1 = tslib_1.__importDefault(require("../abi/poster.json"));
const shares_json_1 = tslib_1.__importDefault(require("../abi/shares.json"));
const superFluidProxy_json_1 = tslib_1.__importDefault(require("../abi/superFluidProxy.json"));
const tributeMinion_json_1 = tslib_1.__importDefault(require("../abi/tributeMinion.json"));
const erc20a_json_1 = tslib_1.__importDefault(require("../abi/erc20a.json"));
const erc721_json_1 = tslib_1.__importDefault(require("../abi/erc721.json"));
const vaultSummoner_json_1 = tslib_1.__importDefault(require("../abi/vaultSummoner.json"));
const gnosisModule_json_1 = tslib_1.__importDefault(require("../abi/gnosisModule.json"));
exports.LOCAL_ABI = {
    BAAL: baal_json_1.default,
    BAAL_SUMMONER: baalSummoner_json_1.default,
    BAAL_ADV_TOKEN_SUMMONER: baalAdvTokenSummoner_json_1.default,
    GNOSIS_MULTISEND: gnosisMultisend_json_1.default,
    GNOSIS_MODULE: gnosisModule_json_1.default,
    GNOSIS_PROXY: gnosisProxy_json_1.default,
    GNOSIS_SIGNLIB: gnosisSignLib_json_1.default,
    LOOT: loot_json_1.default,
    POSTER: poster_json_1.default,
    SHARES: shares_json_1.default,
    SUPERFLUID_PROXY: superFluidProxy_json_1.default,
    TRIBUTE_MINION: tributeMinion_json_1.default,
    ERC20: erc20a_json_1.default,
    ERC721: erc721_json_1.default,
    VAULT_SUMMONER: vaultSummoner_json_1.default,
};
