// @gnosis.pm/zodiac/dist/cjs/types/factories/Integrity__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Integrity__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "NotBFS",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "index",
                type: "uint256",
            },
        ],
        name: "UnsuitableChildCount",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "index",
                type: "uint256",
            },
        ],
        name: "UnsuitableChildTypeTree",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "index",
                type: "uint256",
            },
        ],
        name: "UnsuitableCompValue",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "index",
                type: "uint256",
            },
        ],
        name: "UnsuitableParameterType",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "index",
                type: "uint256",
            },
        ],
        name: "UnsuitableParent",
        type: "error",
    },
    {
        inputs: [],
        name: "UnsuitableRootNode",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "index",
                type: "uint256",
            },
        ],
        name: "UnsupportedOperator",
        type: "error",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint8",
                        name: "parent",
                        type: "uint8",
                    },
                    {
                        internalType: "enum ParameterType",
                        name: "paramType",
                        type: "ParameterType",
                    },
                    {
                        internalType: "enum Operator",
                        name: "operator",
                        type: "Operator",
                    },
                    {
                        internalType: "bytes",
                        name: "compValue",
                        type: "bytes",
                    },
                ],
                internalType: "struct ConditionFlat[]",
                name: "conditions",
                type: "tuple[]",
            },
        ],
        name: "enforce",
        outputs: [],
        stateMutability: "pure",
        type: "function",
    },
];
class Integrity__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.Integrity__factory = Integrity__factory;
Integrity__factory.abi = _abi;
