// @gnosis.pm/zodiac/dist/cjs/types/factories/Packer__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Packer__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint8",
                        name: "parent",
                        type: "uint8",
                    },
                    {
                        internalType: "enum ParameterType",
                        name: "paramType",
                        type: "ParameterType",
                    },
                    {
                        internalType: "enum Operator",
                        name: "operator",
                        type: "Operator",
                    },
                    {
                        internalType: "bytes",
                        name: "compValue",
                        type: "bytes",
                    },
                ],
                internalType: "struct ConditionFlat[]",
                name: "conditionsFlat",
                type: "tuple[]",
            },
        ],
        name: "pack",
        outputs: [
            {
                internalType: "bytes",
                name: "buffer",
                type: "bytes",
            },
        ],
        stateMutability: "pure",
        type: "function",
    },
];
class Packer__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.Packer__factory = Packer__factory;
Packer__factory.abi = _abi;
