// @daohaus/moloch-v3-legos/src/fields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FIELD = void 0;
exports.FIELD = {
    TITLE: {
        id: 'title',
        type: 'input',
        label: '提案標題',
        placeholder: '輸入標題',
    },
    DESCRIPTION: {
        id: 'description',
        type: 'markdownField',
        label: '簡述',
        placeholder: '輸入描述',
    },
    LINK: {
        id: 'link',
        type: 'input',
        label: '連結',
        placeholder: 'http://',
        expectType: 'url',
    },
    METADATA_LINK: {
        id: 'metadataLink',
        type: 'metadataLink',
        label: '連結',
        placeholder: 'http://',
        expectType: 'url',
    },
    SHAMAN_ADDRESS: {
        id: 'shamanAddress',
        type: 'input',
        label: 'Shaman Address',
        placeholder: '0x1234...5678',
    },
    SHAMAN_PERMISSION: {
        id: 'shamanPermission',
        type: 'shamanPermissions',
        label: 'Shaman Permission',
    },
    SHAMAN_DELUXE: {
        id: 'shamanPermission',
        type: 'shamanPermissionDeluxe',
        label: 'Shaman Permission',
    },
    TRIBUTE: {
        id: 'tribute',
        type: 'tributeInput',
        label: 'Tribute',
    },
    REQUEST_TOKEN: {
        id: 'payment',
        type: 'requestERC20',
        label: 'Requested ERC20',
    },
    REQUEST_NATIVE_TOKEN: {
        id: 'paymentAmount',
        type: 'requestNativeToken',
    },
    RAGEQUIT_TOKEN: {
        id: 'sharesToBurn',
        type: 'ragequitToken',
    },
    TO_WEI: {
        id: 'shouldOverwrite',
        type: 'toWeiInput',
        label: 'Should Overwrite',
    },
    NAME: {
        id: 'name',
        type: 'input',
        label: 'DAO 名稱',
        placeholder: 'DAO Name',
    },
    TAGS: {
        id: 'tags',
        type: 'tagsInput',
        label: '標籤 (用逗號分隔)',
        placeholder: 'DAO Name',
    },
    DELEGATE: {
        id: 'delegatingTo',
        type: 'delegateInput',
        label: '代議給',
        placeholder: '0x0...',
    },
    PROPOSAL_EXPIRY: {
        id: 'checkRender',
        type: 'checkRender',
        gateLabel: '增加失效日期',
        components: [
            {
                id: 'proposalExpiry',
                type: 'proposalExpiry',
                defaultValue: '0',
                label: 'Expiration',
            },
        ],
    },
    APPLICANT: {
        id: 'applicantAddress',
        type: 'selectApplicant',
        label: 'Applicant',
        daoMemberOnly: false,
    },
    PROP_OFFERING: {
        id: 'proposalOffering',
        type: 'proposalOffering',
    },
    WALLETCONNECT_LINKER: {
        id: 'walletConnectLink',
        type: 'walletConnectLink',
        label: 'WalletConnect Link',
        info: 'After connecting WalletConnect to another dApp, copy the link and paste it to this input field',
        placeholder: 'wc:',
    },
    SAFE_SELECT: {
        id: 'safeAddress',
        type: 'safeSelect',
        label: 'Safe Name',
        placeholder: 'Select',
    },
    ADDRESSES_AMOUNTS: {
        id: 'addressesAndAmounts',
        type: 'addressesAndAmounts',
        label: 'Addresses & Amounts',
    },
};
