// @daohaus/profile-data/src/profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProfileForAddress = void 0;
const tslib_1 = require("tslib");
const viem_1 = require("viem");
const ens_1 = require("viem/ens");
const chains_1 = require("viem/chains");
const utils_1 = require("./utils");
const getProfileForAddress = ({ address, rpcUri, }) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const reverseEnsRecord = rpcUri && (yield getENSReverseResolver({ address, rpcUri }));
    return (0, utils_1.transformProfile)({
        address,
        ensDomain: reverseEnsRecord,
    });
});
exports.getProfileForAddress = getProfileForAddress;
const getENSReverseResolver = ({ address, rpcUri, }) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const client = (0, viem_1.createPublicClient)({
            chain: chains_1.mainnet,
            transport: (0, viem_1.http)(rpcUri),
        });
        const ensName = yield client.getEnsName({
            address: address,
        });
        if (ensName) {
            const ensText = yield client.getEnsAvatar({
                name: (0, ens_1.normalize)(ensName),
            });
            return {
                domain: {
                    name: ensName,
                    avatar: ensText,
                },
            };
        }
    }
    catch (error) {
        console.error('Unable to fetch ENS reverse record', error);
    }
});
