// viem/chains/dist/cjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formattersOptimism = exports.parseTransactionCelo = exports.serializersCelo = exports.serializeTransactionCelo = exports.formattersCelo = exports.defineTransactionRequest = exports.defineTransactionReceipt = exports.defineTransaction = exports.defineBlock = exports.defineFormatter = exports.defineChain = void 0;
const viem_1 = require("viem");
function defineChain(chain, config = {}) {
    const { fees = chain.fees, formatters = chain.formatters, serializers = chain.serializers, } = config;
    return {
        ...chain,
        fees,
        formatters,
        serializers,
    };
}
exports.defineChain = defineChain;
function defineFormatter(type, format) {
    return ({ exclude, format: overrides, }) => {
        return {
            exclude,
            format: (args) => {
                const formatted = format(args);
                if (exclude) {
                    for (const key of exclude) {
                        delete formatted[key];
                    }
                }
                return {
                    ...formatted,
                    ...overrides(args),
                };
            },
            type,
        };
    };
}
exports.defineFormatter = defineFormatter;
exports.defineBlock = defineFormatter('block', viem_1.formatBlock);
exports.defineTransaction = defineFormatter('transaction', viem_1.formatTransaction);
exports.defineTransactionReceipt = defineFormatter('transactionReceipt', viem_1.formatTransactionReceipt);
exports.defineTransactionRequest = defineFormatter('transactionRequest', viem_1.formatTransactionRequest);
var formatters_js_1 = require("./celo/formatters.js");
Object.defineProperty(exports, "formattersCelo", { enumerable: true, get: function () { return formatters_js_1.formattersCelo; } });
var serializers_js_1 = require("./celo/serializers.js");
Object.defineProperty(exports, "serializeTransactionCelo", { enumerable: true, get: function () { return serializers_js_1.serializeTransactionCelo; } });
Object.defineProperty(exports, "serializersCelo", { enumerable: true, get: function () { return serializers_js_1.serializersCelo; } });
var parsers_js_1 = require("./celo/parsers.js");
Object.defineProperty(exports, "parseTransactionCelo", { enumerable: true, get: function () { return parsers_js_1.parseTransactionCelo; } });
var formatters_js_2 = require("./optimism/formatters.js");
Object.defineProperty(exports, "formattersOptimism", { enumerable: true, get: function () { return formatters_js_2.formattersOptimism; } });
