// @daohaus/utils/src/utils/formatting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isJSON = exports.toBigInt = exports.fromWei = exports.handlePluralNoun = exports.charLimit = exports.truncateAddress = void 0;
const viem_1 = require("viem");
const truncateAddress = (addr) => `${addr.slice(0, 6)}...${addr.slice(-4)}`;
exports.truncateAddress = truncateAddress;
const charLimit = (str, limit) => str && str.length > limit ? `${str.slice(0, limit)}...` : str;
exports.charLimit = charLimit;
const handlePluralNoun = (noun, count) => count === 1 ? noun.singular : noun.plural;
exports.handlePluralNoun = handlePluralNoun;
const fromWei = (amt) => {
    return (0, viem_1.formatEther)(BigInt(amt)).toString();
};
exports.fromWei = fromWei;
const toBigInt = (amt) => {
    if (amt) {
        return BigInt(amt);
    }
};
exports.toBigInt = toBigInt;
const isJSON = (obj) => {
    try {
        JSON.parse(obj);
        return true;
    }
    catch (e) {
        return false;
    }
};
exports.isJSON = isJSON;
