// @shazow/whatsabi/lib.cjs/proxies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.slotResolvers = exports.slots = exports.FixedProxyResolver = exports.SequenceWalletProxyResolver = exports.PROXIABLEProxyResolver = exports.ZeppelinOSProxyResolver = exports.DiamondProxyResolver = exports.EIP1967ProxyResolver = exports.LegacyUpgradeableProxyResolver = exports.GnosisSafeProxyResolver = exports.BaseProxyResolver = void 0;
const utils_js_1 = require("./utils.js");
const _zeroAddress = "0x0000000000000000000000000000000000000000";
function callToAddress(data) {
    return "0x" + data.slice(data.length - 40);
}
class BaseProxyResolver {
    constructor(name) {
        this.name = name || this.constructor.name;
    }
    toString() {
        return this.name;
    }
}
exports.BaseProxyResolver = BaseProxyResolver;
class GnosisSafeProxyResolver extends BaseProxyResolver {
    async resolve(provider, address) {
        const slotPosition = 0;
        return callToAddress(await provider.getStorageAt(address, slotPosition));
    }
}
exports.GnosisSafeProxyResolver = GnosisSafeProxyResolver;
class LegacyUpgradeableProxyResolver extends BaseProxyResolver {
    async resolve(provider, address) {
        const slotPosition = 1;
        return callToAddress(await provider.getStorageAt(address, slotPosition));
    }
}
exports.LegacyUpgradeableProxyResolver = LegacyUpgradeableProxyResolver;
const EIP1967FallbackSelectors = [
    "0x5c60da1b",
    "0xda525716",
    "0xa619486e",
    "0xbb82aa5e",
];
class EIP1967ProxyResolver extends BaseProxyResolver {
    async resolve(provider, address) {
        const implAddr = callToAddress(await provider.getStorageAt(address, exports.slots.EIP1967_IMPL));
        if (implAddr !== _zeroAddress) {
            return implAddr;
        }
        const fallbackAddr = callToAddress(await provider.getStorageAt(address, exports.slots.EIP1967_BEACON));
        if (fallbackAddr === _zeroAddress) {
            return _zeroAddress;
        }
        for (const selector of EIP1967FallbackSelectors) {
            try {
                const addr = callToAddress(await provider.call({
                    to: fallbackAddr,
                    data: selector,
                }));
                if (addr !== _zeroAddress)
                    return addr;
            }
            catch (e) {
                if (e.toString().includes("revert"))
                    continue;
                throw e;
            }
        }
        return _zeroAddress;
    }
}
exports.EIP1967ProxyResolver = EIP1967ProxyResolver;
const diamondSelectors = [
    "0xcdffacc6",
    "0x0d741577",
];
class DiamondProxyResolver extends BaseProxyResolver {
    async resolve(provider, address, selector) {
        if (!selector) {
            throw "DiamondProxy requires a selector to resolve to a specific facet";
        }
        else if (selector.startsWith("0x")) {
            selector = selector.slice(2);
        }
        const facetMappingSlot = (0, utils_js_1.keccak256)("0x" + selector.padEnd(64, "0") + exports.slots.DIAMOND_STORAGE.slice(2));
        const facet = await provider.getStorageAt(address, facetMappingSlot);
        const storageAddr = "0x" + facet.slice(facet.length - 40);
        if (storageAddr !== _zeroAddress) {
            return storageAddr;
        }
        for (const facetSelector of diamondSelectors) {
            try {
                const addr = callToAddress(await provider.call({
                    to: address,
                    data: facetSelector + selector,
                }));
                if (addr !== _zeroAddress)
                    return addr;
            }
            catch (e) {
                if (e.toString().includes("revert"))
                    continue;
                throw e;
            }
        }
        return _zeroAddress;
    }
}
exports.DiamondProxyResolver = DiamondProxyResolver;
class ZeppelinOSProxyResolver extends BaseProxyResolver {
    async resolve(provider, address) {
        return callToAddress(await provider.getStorageAt(address, exports.slots.ZEPPELINOS_IMPL));
    }
}
exports.ZeppelinOSProxyResolver = ZeppelinOSProxyResolver;
class PROXIABLEProxyResolver extends BaseProxyResolver {
    async resolve(provider, address) {
        return callToAddress(await provider.getStorageAt(address, exports.slots.PROXIABLE));
    }
}
exports.PROXIABLEProxyResolver = PROXIABLEProxyResolver;
class SequenceWalletProxyResolver extends BaseProxyResolver {
    async resolve(provider, address) {
        return callToAddress(await provider.getStorageAt(address, address.toLowerCase().slice(2)));
    }
    toString() {
        return "SequenceWalletProxy";
    }
}
exports.SequenceWalletProxyResolver = SequenceWalletProxyResolver;
class FixedProxyResolver extends BaseProxyResolver {
    constructor(name, resolvedAddress) {
        super(name);
        this.resolvedAddress = resolvedAddress;
    }
    async resolve(provider, address) {
        return this.resolvedAddress;
    }
}
exports.FixedProxyResolver = FixedProxyResolver;
;
exports.slots = {
    EIP1967_IMPL: "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
    EIP1967_BEACON: "0xa3f0ad74e5423aebfd80d3ef4346578335a9a72aeaee59ff6cb3582b35133d50",
    ZEPPELINOS_IMPL: "0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3",
    PROXIABLE: "0xc5f16f0fcc639fa48a6947836d9850f504798523bf8c9a3a87d5876cf622bcf7",
    GNOSIS_SAFE_SELECTOR: "0xa619486e00000000000000000000000000000000000000000000000000000000",
    DIAMOND_STORAGE: "0xc8fcad8db84d3cc18b4c41d551ea0ee66dd599cde068d998e57d5e09332c131b",
};
exports.slotResolvers = {
    [exports.slots.EIP1967_IMPL]: new EIP1967ProxyResolver("EIP1967Proxy"),
    [exports.slots.EIP1967_BEACON]: new EIP1967ProxyResolver("EIP1967Proxy"),
    [exports.slots.ZEPPELINOS_IMPL]: new ZeppelinOSProxyResolver("ZeppelinOSProxy"),
    [exports.slots.PROXIABLE]: new PROXIABLEProxyResolver("PROXIABLE"),
    [exports.slots.GNOSIS_SAFE_SELECTOR]: new GnosisSafeProxyResolver("GnosisSafeProxy"),
    [exports.slots.DIAMOND_STORAGE]: new DiamondProxyResolver("DiamondProxy"),
    ["0xc8fcad8db84d3cc18b4c41d551ea0ee66dd599cde068d998e57d5e09332c131d"]: new DiamondProxyResolver("DiamondProxy"),
};
