// ethers-multisend/build/esm/createTransaction.js
import { TransactionType } from './types';
export const createTransaction = (type, id = '') => {
    switch (type) {
        case TransactionType.callContract:
            return {
                type,
                id,
                to: '',
                value: '',
                abi: '',
                functionSignature: '',
                inputValues: {},
            };
        case TransactionType.transferFunds:
            return {
                type,
                id,
                token: '',
                to: '',
                amount: '',
                decimals: 18,
            };
        case TransactionType.transferCollectible:
            return {
                type,
                id,
                address: '',
                tokenId: '',
                to: '',
                from: '',
            };
        case TransactionType.raw:
            return { type, id, to: '', value: '', data: '' };
    }
    throw new Error(`Invalid transaction type: ${type}`);
};
