// @gnosis.pm/zodiac/dist/esm/types/factories/MultisendEncoder__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [],
        name: "NoTransactions",
        type: "error",
    },
    {
        inputs: [],
        name: "UnequalArraysLengths",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "multisend",
                type: "address",
            },
            {
                internalType: "address[]",
                name: "targets",
                type: "address[]",
            },
            {
                internalType: "uint256[]",
                name: "values",
                type: "uint256[]",
            },
            {
                internalType: "bytes[]",
                name: "calldatas",
                type: "bytes[]",
            },
        ],
        name: "encodeMultisend",
        outputs: [
            {
                internalType: "address",
                name: "to",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "value",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
            {
                internalType: "enum Enum.Operation",
                name: "operation",
                type: "Enum.Operation",
            },
        ],
        stateMutability: "pure",
        type: "function",
    },
];
export class MultisendEncoder__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
MultisendEncoder__factory.abi = _abi;
