// @gnosis.pm/zodiac/dist/esm/types/factories/Packer__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint8",
                        name: "parent",
                        type: "uint8",
                    },
                    {
                        internalType: "enum ParameterType",
                        name: "paramType",
                        type: "ParameterType",
                    },
                    {
                        internalType: "enum Operator",
                        name: "operator",
                        type: "Operator",
                    },
                    {
                        internalType: "bytes",
                        name: "compValue",
                        type: "bytes",
                    },
                ],
                internalType: "struct ConditionFlat[]",
                name: "conditionsFlat",
                type: "tuple[]",
            },
        ],
        name: "pack",
        outputs: [
            {
                internalType: "bytes",
                name: "buffer",
                type: "bytes",
            },
        ],
        stateMutability: "pure",
        type: "function",
    },
];
export class Packer__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
Packer__factory.abi = _abi;
