// @daohaus/keychain-utils/src/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ABI_EXPLORER_KEYS = exports.PINATA_API_KEYS = exports.GRAPH_API_KEYS = exports.HAUS_RPC = exports.HAUS_RPC_DEFAULTS = exports.getGraphUrl = exports.addApiKeyToGraphEnpoints = exports.ENDPOINTS = void 0;
exports.ENDPOINTS = {
    V3_SUBGRAPH: {
        '0x1': 'https://gateway.thegraph.com/api/[api-key]/subgraphs/id/GfHFdFmiSwW1PKtnDhhcxhArwtTjVuMnXxQ5XcETF1bP',
        '0x5': 'https://api.thegraph.com/subgraphs/name/hausdao/daohaus-v3-goerli',
        '0x64': 'https://gateway.thegraph.com/api/[api-key]/subgraphs/id/5oXHJGgAWTSEXHK5FGf7mcxYtk6rSz7MJV5ccvGfoW6q',
        '0x89': 'https://api.thegraph.com/subgraphs/name/hausdao/daohaus-v3-polygon',
        '0xa': 'https://api.thegraph.com/subgraphs/name/hausdao/daohaus-v3-optimism',
        '0xa4b1': 'https://api.thegraph.com/subgraphs/name/hausdao/daohaus-v3-arbitrum',
    },
    EXPLORER: {
        '0x1': 'https://etherscan.io',
        '0x5': 'https://goerli.etherscan.io',
        '0x64': 'https://gnosisscan.io',
        '0x89': 'https://polygonscan.com',
        '0xa': 'https://optimistic.etherscan.io',
        '0xa4b1': 'https://arbiscan.io',
    },
    GNOSIS_API: {
        '0x1': 'https://safe-transaction-mainnet.safe.global/api/v1',
        '0x5': 'https://safe-transaction-goerli.safe.global/api/v1',
        '0x64': 'https://safe-transaction-gnosis-chain.safe.global/api/v1',
        '0x89': 'https://safe-transaction-polygon.safe.global/api/v1',
        '0xa': 'https://safe-transaction-optimism.safe.global/api/v1',
        '0xa4b1': 'https://safe-transaction-arbitrum.safe.global/api/v1',
    },
    GNOSIS_SAFE_UI: {
        '0x1': 'https://app.safe.global/eth',
        '0x5': 'https://app.safe.global/gor',
        '0x64': 'https://app.safe.global/gno',
        '0x89': 'https://app.safe.global/matic',
        '0xa': 'https://app.safe.global/oeth',
        '0xa4b1': 'https://app.safe.global/arb',
    },
    TABULA_GRAPH: {
        '0x5': 'https://api.thegraph.com/subgraphs/name/auryn-macmillan/tabula-goerli',
    },
};
const addApiKeyToGraphEnpoints = (graphApiKeys, endpoints) => {
    return Object.keys(graphApiKeys).reduce((acc, key) => {
        if (endpoints['V3_SUBGRAPH'][key] && acc) {
            const unreplacedValue = acc['V3_SUBGRAPH'][key];
            const apiKey = graphApiKeys[key];
            if (unreplacedValue && apiKey) {
                acc['V3_SUBGRAPH'][key] = unreplacedValue.replace('[api-key]', apiKey);
            }
        }
        return acc;
    }, endpoints);
};
exports.addApiKeyToGraphEnpoints = addApiKeyToGraphEnpoints;
const getGraphUrl = (networkId, graphApiKeys) => {
    if (graphApiKeys) {
        const endpoints = (0, exports.addApiKeyToGraphEnpoints)(graphApiKeys, exports.ENDPOINTS);
        return endpoints['V3_SUBGRAPH'][networkId];
    }
    return exports.ENDPOINTS['V3_SUBGRAPH'][networkId];
};
exports.getGraphUrl = getGraphUrl;
exports.HAUS_RPC_DEFAULTS = {
    '0x1': process.env['NX_RIVET_KEY']
        ? `https://${process.env['NX_RIVET_KEY']}.eth.rpc.rivet.cloud`
        : `https://eth-mainnet.g.alchemy.com/v2/${process.env['NX_ETHEREUM_ALCHEMY_KEY']}`,
    '0x5': process.env['NX_RIVET_KEY']
        ? `https://${process.env['NX_RIVET_KEY']}.goerli.rpc.rivet.cloud/`
        : `https://eth-goerli.g.alchemy.com/v2/demo`,
    '0x64': 'https://rpc.gnosischain.com/',
    '0xa': process.env['NX_OPTIMISM_ALCHEMY_KEY']
        ? `https://opt-mainnet.g.alchemy.com/v2/${process.env['NX_OPTIMISM_ALCHEMY_KEY']}`
        : 'https://mainnet.optimism.io',
    '0x89': process.env['NX_POLYGONPOS_ALCHEMY_KEY']
        ? `https://polygon-mainnet.g.alchemy.com/v2/${process.env['NX_POLYGONPOS_ALCHEMY_KEY']}`
        : 'https://polygon-rpc.com/',
    '0xa4b1': process.env['NX_ARBITRUM_ALCHEMY_KEY']
        ? `https://arb-mainnet.g.alchemy.com/v2/${process.env['NX_ARBITRUM_ALCHEMY_KEY']}`
        : 'https://arb1.arbitrum.io/rpc',
};
exports.HAUS_RPC = {
    '0x1': process.env['NX_MAINNET_RPC']
        ? process.env['NX_MAINNET_RPC']
        : exports.HAUS_RPC_DEFAULTS['0x1'],
    '0x5': process.env['NX_GOERLI_RPC']
        ? process.env['NX_GOERLI_RPC']
        : exports.HAUS_RPC_DEFAULTS['0x5'],
    '0x64': process.env['NX_GNOSISCHAIN_RPC']
        ? process.env['NX_GNOSISCHAIN_RPC']
        : exports.HAUS_RPC_DEFAULTS['0x64'],
    '0xa': process.env['NX_OPTIMISM_RPC']
        ? process.env['NX_OPTIMISM_RPC']
        : exports.HAUS_RPC_DEFAULTS['0xa'],
    '0x89': process.env['NX_POLYGONPOS_RPC']
        ? process.env['NX_POLYGONPOS_RPC']
        : exports.HAUS_RPC_DEFAULTS['0x89'],
    '0xa4b1': process.env['NX_ARBITRUM_RPC']
        ? process.env['NX_ARBITRUM_RPC']
        : exports.HAUS_RPC_DEFAULTS['0xa4b1'],
};
exports.GRAPH_API_KEYS = {
    '0x1': process.env['NX_GRAPH_API_KEY_MAINNET'],
    '0x64': process.env['NX_GRAPH_API_KEY_MAINNET'],
};
exports.PINATA_API_KEYS = {
    pinata_api_key: process.env['NX_PINATA_API_KEY'],
    pinata_api_secret: process.env['NX_PINATA_API_SECRET'],
};
exports.ABI_EXPLORER_KEYS = {
    '0x1': process.env['NX_ETHERSCAN_KEY'],
    '0x5': process.env['NX_ETHERSCAN_KEY'],
    '0x64': process.env['NX_GNOSISSCAN_KEY'],
    '0x89': process.env['NX_POLYGONSCAN_KEY'],
    '0xa': process.env['NX_OPTIMISMSCAN_KEY'],
    '0xa4b1': process.env['NX_ARBISCAN_KEY'],
};
