// @daohaus/keychain-utils/src/networkData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HAUS_NETWORK_DATA = void 0;
const endpoints_1 = require("./endpoints");
exports.HAUS_NETWORK_DATA = {
    '0x1': {
        chainId: '0x1',
        networkId: 1,
        name: 'Mainnet',
        symbol: 'ETH',
        tokenDecimals: 18,
        explorer: endpoints_1.ENDPOINTS.EXPLORER['0x1'],
        rpc: endpoints_1.HAUS_RPC['0x1'],
    },
    '0x5': {
        chainId: '0x5',
        networkId: 5,
        name: 'Goerli',
        symbol: 'ETH',
        tokenDecimals: 18,
        explorer: endpoints_1.ENDPOINTS.EXPLORER['0x5'],
        rpc: endpoints_1.HAUS_RPC['0x5'],
    },
    '0x64': {
        chainId: '0x64',
        networkId: 100,
        name: 'Gnosis Chain',
        symbol: 'XDAI',
        tokenDecimals: 18,
        explorer: endpoints_1.ENDPOINTS.EXPLORER['0x64'],
        rpc: endpoints_1.HAUS_RPC['0x64'],
    },
    '0x89': {
        chainId: '0x89',
        networkId: 137,
        name: 'Polygon',
        symbol: 'MATIC',
        tokenDecimals: 18,
        explorer: endpoints_1.ENDPOINTS.EXPLORER['0x89'],
        rpc: endpoints_1.HAUS_RPC['0x89'],
    },
    '0xa': {
        chainId: '0xa',
        networkId: 10,
        name: 'Optimism',
        symbol: 'ETH',
        tokenDecimals: 18,
        explorer: endpoints_1.ENDPOINTS.EXPLORER['0xa'],
        rpc: endpoints_1.HAUS_RPC['0xa'],
    },
    '0xa4b1': {
        chainId: '0xa4b1',
        networkId: 42161,
        name: 'Arbitrum',
        symbol: 'ETH',
        tokenDecimals: 18,
        explorer: endpoints_1.ENDPOINTS.EXPLORER['0xa4b1'],
        rpc: endpoints_1.HAUS_RPC['0xa4b1'],
    },
};
