// ethers-multisend/build/cjs/createTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTransaction = void 0;
const types_1 = require("./types");
const createTransaction = (type, id = '') => {
    switch (type) {
        case types_1.TransactionType.callContract:
            return {
                type,
                id,
                to: '',
                value: '',
                abi: '',
                functionSignature: '',
                inputValues: {},
            };
        case types_1.TransactionType.transferFunds:
            return {
                type,
                id,
                token: '',
                to: '',
                amount: '',
                decimals: 18,
            };
        case types_1.TransactionType.transferCollectible:
            return {
                type,
                id,
                address: '',
                tokenId: '',
                to: '',
                from: '',
            };
        case types_1.TransactionType.raw:
            return { type, id, to: '', value: '', data: '' };
    }
    throw new Error(`Invalid transaction type: ${type}`);
};
exports.createTransaction = createTransaction;
