// ethers-multisend/build/cjs/decodeSingle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeSingle = void 0;
const abi_1 = require("@ethersproject/abi");
const bignumber_1 = require("@ethersproject/bignumber");
const contracts_1 = require("@ethersproject/contracts");
const units_1 = require("@ethersproject/units");
const interfaces_1 = require("./interfaces");
const types_1 = require("./types");
const decodeSingle = async (transaction, provider, fetchAbi, id = '') => {
    const { to, data, value } = transaction;
    if (!data || bignumber_1.BigNumber.from(data).isZero()) {
        // ETH transfer
        return {
            type: types_1.TransactionType.transferFunds,
            id,
            to,
            amount: (0, units_1.formatEther)(value).toString(),
            decimals: 18,
            token: null,
        };
    }
    let erc20TransferData = null;
    try {
        erc20TransferData = interfaces_1.erc20Interface.decodeFunctionData(interfaces_1.erc20TransferFragment, data);
    }
    catch (e) {
        // it's not an ERC20 transfer
    }
    if (erc20TransferData && bignumber_1.BigNumber.from(value).isZero()) {
        const decimals = await new contracts_1.Contract(to, interfaces_1.erc20Interface, provider).decimals();
        return {
            type: types_1.TransactionType.transferFunds,
            id,
            to: erc20TransferData.recipient,
            amount: (0, bignumber_1.formatFixed)(erc20TransferData.amount, decimals),
            decimals,
            token: to,
        };
    }
    let erc721TransferData = null;
    try {
        erc721TransferData = interfaces_1.erc721Interface.decodeFunctionData(interfaces_1.erc721TransferFragment, data);
    }
    catch (e) {
        // it's not an ERC721 transfer
    }
    if (erc721TransferData && bignumber_1.BigNumber.from(value).isZero()) {
        return {
            type: types_1.TransactionType.transferCollectible,
            id,
            from: erc721TransferData._from,
            to: erc721TransferData._to,
            tokenId: erc721TransferData._tokenId.toString(),
            address: to,
        };
    }
    // try to fetch the ABI
    let abi = '';
    if (fetchAbi) {
        try {
            abi = (await fetchAbi(to, data)) || '';
        }
        catch (e) {
            // could not fetch ABI
        }
    }
    if (abi) {
        const iface = new abi_1.Interface(abi);
        let fragment;
        try {
            fragment = iface.getFunction(data.substring(0, 10).toLowerCase());
        }
        catch (e) {
            // ignore error if fragment is not found in abi
        }
        if (fragment) {
            return {
                type: types_1.TransactionType.callContract,
                id,
                to,
                abi,
                functionSignature: fragment.format(),
                inputValues: decodeArgs(data, fragment.inputs),
                value: bignumber_1.BigNumber.from(value || '0x00').toString(),
            };
        }
    }
    return {
        type: types_1.TransactionType.raw,
        id,
        to,
        value: bignumber_1.BigNumber.from(value || '0x00').toString(),
        data,
    };
};
exports.decodeSingle = decodeSingle;
// we slightly adjust ethers' default coerce function so we return BigNumbers as strings.
const abiCoder = new abi_1.AbiCoder((name, value) => {
    if (!bignumber_1.BigNumber.isBigNumber(value))
        return value;
    // Return as number if not too big.
    // (This is replicating the ethers' default coerce function.)
    const match = name.match('^u?int([0-9]+)$');
    if (match && parseInt(match[1]) <= 48) {
        return value.toNumber();
    }
    return value.toString();
});
const decodeArgs = (data, inputs) => {
    const result = abiCoder.decode(inputs, '0x' + data.substring(10));
    const keys = Object.keys(result);
    const namedKeys = keys.filter((key) => `${parseInt(key)}` !== key);
    const allArgsHaveNames = namedKeys.length * 2 === keys.length;
    const keysToUse = allArgsHaveNames ? namedKeys : keys;
    return Object.assign({}, ...keysToUse.map((key) => ({ [key]: result[key] })));
};
