// ethers-multisend/build/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValid = exports.createTransaction = exports.decodeSingle = exports.encodeSingle = exports.decodeMulti = exports.encodeMulti = void 0;
var encodeMulti_1 = require("./encodeMulti");
Object.defineProperty(exports, "encodeMulti", { enumerable: true, get: function () { return encodeMulti_1.encodeMulti; } });
var decodeMulti_1 = require("./decodeMulti");
Object.defineProperty(exports, "decodeMulti", { enumerable: true, get: function () { return decodeMulti_1.decodeMulti; } });
var encodeSingle_1 = require("./encodeSingle");
Object.defineProperty(exports, "encodeSingle", { enumerable: true, get: function () { return encodeSingle_1.encodeSingle; } });
var decodeSingle_1 = require("./decodeSingle");
Object.defineProperty(exports, "decodeSingle", { enumerable: true, get: function () { return decodeSingle_1.decodeSingle; } });
var createTransaction_1 = require("./createTransaction");
Object.defineProperty(exports, "createTransaction", { enumerable: true, get: function () { return createTransaction_1.createTransaction; } });
var isValid_1 = require("./isValid");
Object.defineProperty(exports, "isValid", { enumerable: true, get: function () { return isValid_1.isValid; } });
__exportStar(require("./types"), exports);
