// @daohaus/moloch-v3-data/src/utils/paging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPaging = exports.paginateResponse = exports.DEFAULT_RECORDS_PER_PAGE = void 0;
exports.DEFAULT_RECORDS_PER_PAGE = 25;
const paginateResponse = (items, paging) => {
    const hasNextPage = paging.pageSize < items.length;
    return hasNextPage ? items.slice(0, paging.pageSize) : items;
};
exports.paginateResponse = paginateResponse;
const createPaging = (items, paging) => {
    const hasNextPage = paging.pageSize < items.length;
    const pageItems = hasNextPage ? items.slice(0, paging.pageSize) : items;
    return {
        pageItems,
        nextPaging: hasNextPage
            ? {
                pageSize: paging.pageSize,
                offset: getNextOffset(paging),
                lastId: getNextLastId(paging, pageItems),
            }
            : undefined,
        previousPaging: hasPreviousOffsetPage(paging)
            ? {
                pageSize: paging.pageSize,
                offset: getPreviousOffset(paging),
            }
            : undefined,
    };
};
exports.createPaging = createPaging;
const hasPreviousOffsetPage = (paging) => paging.offset !== undefined && paging.offset > 0;
const getNextOffset = (paging) => {
    return paging.offset !== undefined
        ? paging.offset + paging.pageSize
        : paging.offset;
};
const getPreviousOffset = (paging) => {
    if (paging.offset !== undefined) {
        return paging.offset - paging.pageSize;
    }
    return paging.offset;
};
const getNextLastId = (paging, pageItems) => {
    var _a;
    return paging.lastId !== undefined ? (_a = pageItems.slice(-1)[0]) === null || _a === void 0 ? void 0 : _a.id : undefined;
};
