// @daohaus/baal-contracts/export/src/addresses/setup.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSetupAddresses = void 0;
var ethers_1 = require("ethers");
var getSetupAddresses = function (chainId, network, deployments) { return __awaiter(void 0, void 0, void 0, function () {
    var _a;
    var _b, _c, _d, _e, _f;
    return __generator(this, function (_g) {
        switch (_g.label) {
            case 0:
                if (!(network.name === 'hardhat')) return [3 /*break*/, 7];
                if (!deployments)
                    throw Error("Must specify contract deployments for ".concat(network.name, " network"));
                _a = {};
                return [4 /*yield*/, deployments.get('GnosisSafe')];
            case 1:
                _a.gnosisSingleton = (_b = (_g.sent())) === null || _b === void 0 ? void 0 : _b.address;
                return [4 /*yield*/, deployments.get('CompatibilityFallbackHandler')];
            case 2:
                _a.gnosisFallbackLibrary = (_c = (_g.sent())) === null || _c === void 0 ? void 0 : _c.address;
                return [4 /*yield*/, deployments.get('MultiSend')];
            case 3:
                _a.gnosisMultisendLibrary = (_d = (_g.sent())) === null || _d === void 0 ? void 0 : _d.address;
                return [4 /*yield*/, (deployments === null || deployments === void 0 ? void 0 : deployments.get('Poster'))];
            case 4:
                _a.poster = (_g.sent()).address;
                return [4 /*yield*/, deployments.get('GnosisSafeProxyFactory')];
            case 5:
                _a.gnosisSafeProxyFactory = (_e = (_g.sent())) === null || _e === void 0 ? void 0 : _e.address;
                return [4 /*yield*/, deployments.get('ModuleProxyFactory')];
            case 6: return [2 /*return*/, (_a.moduleProxyFactory = (_f = (_g.sent())) === null || _f === void 0 ? void 0 : _f.address,
                    _a.DAO = ethers_1.constants.AddressZero,
                    _a)];
            case 7:
                // same default for all networks, but different sometimes
                // https://github.com/gnosis/zodiac/blob/master/src/factory/constants.ts#L20
                // https://github.com/safe-global/safe-deployments/tree/main/src/assets
                // moduleProxyFactory https://github.com/gnosis/zodiac/blob/master/src/factory/constants.ts#L21
                if (chainId !== '10') { // Optimism
                    return [2 /*return*/, {
                            gnosisSingleton: "0xd9db270c1b5e3bd161e8c8503c55ceabee709552",
                            gnosisFallbackLibrary: "0xf48f2b2d2a534e402487b3ee7c18c33aec0fe5e4",
                            gnosisMultisendLibrary: "0xa238cbeb142c10ef7ad8442c6d1f9e89e07e7761",
                            poster: "0x000000000000cd17345801aa8147b8D3950260FF",
                            gnosisSafeProxyFactory: "0xa6B71E26C5e0845f74c812102Ca7114b6a896AB2",
                            moduleProxyFactory: "0x00000000000DC7F163742Eb4aBEf650037b1f588",
                            DAO: "0x4A9a27d614a74Ee5524909cA27bdBcBB7eD3b315" // Change to Daohaus protocol zodiac baal avatar
                        }];
                }
                // Optimism ONLY
                // Safe singleton: 0x69f4D1788e39c87893C980c06EdF4b7f686e2938 
                // fall back: 0x017062a1dE2FE6b99BE3d9d37841FeD19F573804	
                // multisend: 0x998739BFdAAdde7C933B942a68053933098f9EDa
                // GnosisSafeProxyFactory: 0xC22834581EbC8527d974F8a1c97E1bEA4EF910BC
                // DAO: "0x1aCFF11474B9C6D15966Da7A08eD23438CDE23D4"
                return [2 /*return*/, {
                        gnosisSingleton: "0x69f4D1788e39c87893C980c06EdF4b7f686e2938",
                        gnosisFallbackLibrary: "0x017062a1dE2FE6b99BE3d9d37841FeD19F573804",
                        gnosisMultisendLibrary: "0x998739BFdAAdde7C933B942a68053933098f9EDa",
                        poster: "0x000000000000cd17345801aa8147b8D3950260FF",
                        gnosisSafeProxyFactory: "0xC22834581EbC8527d974F8a1c97E1bEA4EF910BC",
                        moduleProxyFactory: "0x00000000000DC7F163742Eb4aBEf650037b1f588",
                        DAO: "0x1aCFF11474B9C6D15966Da7A08eD23438CDE23D4" // Change to Daohaus protocol zodiac baal avatar
                    }];
        }
    });
}); };
exports.getSetupAddresses = getSetupAddresses;
